src/plugins/api-network.ts

```diff
@@ -0,0 +1,156 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+
+import { Frame, Page, Request } from 'playwright-core';
+import { NetworkInfo, APIDriver } from '../common_types';
+import { log } from '../core/logger';
+import { Step } from '../dsl';
+import { getTimestamp } from '../helpers';
+
+/**
+ * Kibana UI expects the requestStartTime and loadEndTime to be baseline
+ * in seconds as they have the logic to convert it to milliseconds before
+ * using for offset calculation
+ */
+function epochTimeInSeconds() {
+  return getTimestamp() / 1e6;
+}
+
+const roundMilliSecs = (value: number): number => {
+  return Math.floor(value * 1000) / 1000;
+};
+
+export class APINetworkManager {
+  private _barrierPromises = new Set<Promise<void>>();
+  results: Array<NetworkInfo> = [];
+  _currentStep: Partial<Step> = null;
+  _originalMethods: any;
+
+  constructor(private driver: APIDriver) {
+    this._originalMethods = {};
+  }
+
+  /**
+   * Adds a protection barrier aganist all asynchronous extract operations from
+   * request/response object that are happening during page lifecycle, If the
+   * page is closed during the extraction, the barrier enforces those operations
+   * to not result in exception
+   */
+  private _addBarrier(page: Page, promise: Promise<void>) {
+    if (!page) return;
+    const race = Promise.race([
+      new Promise<void>(resolve =>
+        page.on('close', () => {
+          this._barrierPromises.delete(race);
+          resolve();
+        })
+      ),
+      promise,
+    ]);
+    this._barrierPromises.add(race);
+    race.then(() => this._barrierPromises.delete(race));
+  }
+
+  private _nullableFrameBarrier(req: Request): Frame | null {
+    try {
+      return req.frame();
+    } catch (_) {
+      // frame might be unavailable for certain requests if they are issued
+      // before the frame is created - true for navigation requests
+      // https://playwright.dev/docs/api/class-request#request-frame
+    }
+    return null;
+  }
+
+  async start() {
+    log(`Plugins: started collecting network events`);
+    const { request } = this.driver;
+
+    // Intercept API requests
+    ['fetch', 'get', 'post', 'put', 'delete', 'patch', 'head'].forEach(
+      method => {
+        this._originalMethods[method] = request[method].bind(request);
+        request[method] = this._interceptRequest.bind(this, method);
+      }
+    );
+  }
+
+  async _interceptRequest(method, url, options: any) {
```

Comment:
This needs some careful review as we need some extensive tests for cached, aborted, intercepted requests etc, Will do a detailed review later on when we have tests etc.